typegate/src/typegate/artifacts/shared.ts

```diff
@@ -0,0 +1,255 @@
+// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+
+import { connect, Redis, RedisConnectOptions } from "redis";
+import { createHash } from "node:crypto";
+import { S3 } from "aws-sdk/client-s3";
+import { getLocalPath, STORE_DIR, STORE_TEMP_DIR } from "./mod.ts";
+import { ArtifactMeta, ArtifactStore } from "./mod.ts";
+import { HashTransformStream } from "../../utils/hash.ts";
+import { resolve } from "std/path/resolve.ts";
+import { SyncConfig } from "../../sync/config.ts";
+import { readAll } from "https://deno.land/std@0.129.0/streams/conversion.ts";
+
+export interface RemoteUploadUrlStore {
+  redisClient: Redis;
+}
+
+const setCmd = `
+local key = KEYS[1]
+local value = ARGV[1]
+local expirationTime = ARGV[2]
+
+redis.call('HSET', key, 'url', value)
+redis.call('EXPIRE', key, expirationTime)
+`.trim();
+const existsCmd = `
+local key = KEYS[1]
+
+local exists = redis.call('EXISTS', key)
+return exists
+`.trim();
+
+function resolveRedisUrlKey(url: string) {
+  return `articact-upload-urls:${url}`;
+}
+
+async function initRemoteUploadUrlStore(
+  redisConfig: RedisConnectOptions,
+): Promise<RemoteUploadUrlStore> {
+  const redisClient = await connect(redisConfig);
+
+  return { redisClient };
+}
+
+async function deinitRemoteUploadUrlStore(uploadUrls: RemoteUploadUrlStore) {
+  await uploadUrls.redisClient.quit();
+}
+
+function serializeToRedisValue<T>(value: T): string {
+  return JSON.stringify(value);
+}
+
+function deserializeToCustom<T>(value: string): T {
+  try {
+    return JSON.parse(value) as T;
+  } catch (error) {
+    throw error;
```

Comment:
isn't this `try...catch` redundant?

Reply:
I haven't handled error related to converting the parsed value to the required Type anywhere else

Reply:
catch / throw is a no-op, is that really wanted?

Reply:
Oww, OK, understood it now. Will change it.

Reply:
Done.